File: server/src/routes/outlier.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const { v4: uuidv4 } = require('uuid');
const ExcelJS = require('exceljs');
const outlierDetectionService = require('../services/outlierDetectionService');
const queueService = require('../services/queueService');
const { authenticate, requireScopes } = require('../middleware/auth');
const { requirePermission, requireOwnershipOrAdmin } = require('../middleware/rbac');
const { getAnalysisRepository } = require('../repositories');
const logger = require('../utils/logger');

const router = express.Router();

// Get repository instance
let analysisRepository;
getAnalysisRepository().then(repo => {
  analysisRepository = repo;
}).catch(error => {
  logger.error('Error initializing analysis repository:', error);
});

// Start new outlier analysis
router.post('/start',
  authenticate,
  requirePermission('analysis:write'),
  requireScopes(['write']), // For API key access
  [
    body('exclusionChannels').isArray().withMessage('Exclusion channels must be an array'),
    body('minSubs').isInt({ min: 1000 }).withMessage('Minimum subscribers must be at least 1000'),
    body('maxSubs').isInt({ min: 10000 }).withMessage('Maximum subscribers must be at least 10000'),
    body('timeWindow').isInt({ min: 1, max: 30 }).withMessage('Time window must be between 1-30 days'),
    body('outlierThreshold').isInt({ min: 10, max: 100 }).withMessage('Outlier threshold must be between 10-100'),
  ], async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({
          success: false,
          errors: errors.array(),
        });
      }

      const analysisId = uuidv4();
      const config = req.body;
      const io = req.app.get('io');

      // Create analysis record in database
      const analysisData = {
        id: analysisId,
        user_id: req.user?.id || null,
        name: config.name || `Analysis ${new Date().toISOString()}`,
        description: config.description || null,
        config,
        status: 'pending',
      };

      try {
        await analysisRepository.createAnalysis(analysisData);
      } catch (error) {
        logger.error('Error creating analysis record:', error);
        return res.status(500).json({
          success: false,
          message: 'Failed to create analysis record',
        });
      }

      // Determine if we should use queue system or direct processing
      const useQueue = process.env.USE_QUEUE_SYSTEM === 'true' || config.useQueue;
      
      if (useQueue) {
        try {
          // Use queue system for background processing
          const job = await queueService.startYouTubeAnalysis(
            analysisId,
            config,
            req.user?.id,
            {
              priority: config.priority || 75,
              socketRoom: `analysis-${analysisId}`,
            }
          );
          
          logger.info(`Analysis ${analysisId} queued with job ID: ${job.jobId}`);
        } catch (queueError) {
          logger.warn(`Failed to queue analysis, falling back to direct processing:`, queueError);
          
          // Fallback to direct processing
          outlierDetectionService.startAnalysis(analysisId, config, io)
            .then(results => {
              logger.info(`Analysis ${analysisId} completed successfully (direct)`);
            })
            .catch(error => {
              logger.error(`Analysis ${analysisId} failed (direct):`, error);
            });
        }
      } else {
        // Use direct processing (original behavior)
        outlierDetectionService.startAnalysis(analysisId, config, io)
          .then(results => {
            logger.info(`Analysis ${analysisId} completed successfully (direct)`);
          })
          .catch(error => {
            logger.error(`Analysis ${analysisId} failed (direct):`, error);
          });
      }

      res.json({
        success: true,
        analysisId,
        message: 'Analysis started successfully',
      });

    } catch (error) {
      next(error);
    }
  });

// Get analysis status
router.get('/status/:analysisId',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res) => {
    try {
      const { analysisId } = req.params;
      const analysis = await analysisRepository.findById(analysisId);

      if (!analysis) {
        return res.status(404).json({
          success: false,
          message: 'Analysis not found',
        });
      }

      // Check ownership (users can only access their own analyses, admins can access all)
      if (analysis.user_id !== req.user.id && req.user.role !== 'admin') {
        return res.status(403).json({
          success: false,
          message: 'Access denied',
        });
      }

      res.json({
        success: true,
        analysis: {
          id: analysis.id,
          status: analysis.status,
          started_at: analysis.started_at,
          completed_at: analysis.completed_at,
          config: analysis.config,
          processing_time_ms: analysis.processing_time_ms,
          total_outliers_found: analysis.total_outliers_found,
          total_channels_analyzed: analysis.total_channels_analyzed,
          error_message: analysis.error_message,
        },
      });
    } catch (error) {
      logger.error('Error fetching analysis status:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error',
      });
    }
  });

// Get analysis results
router.get('/results/:analysisId',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res) => {
    try {
      const { analysisId } = req.params;
      const analysis = await analysisRepository.findById(analysisId);

      if (!analysis) {
        return res.status(404).json({
          success: false,
          message: 'Analysis not found',
        });
      }

      // Check ownership (users can only access their own analyses, admins can access all)
      if (analysis.user_id !== req.user.id && req.user.role !== 'admin') {
        return res.status(403).json({
          success: false,
          message: 'Access denied',
        });
      }

      if (analysis.status !== 'completed') {
        return res.status(400).json({
          success: false,
          message: `Analysis is ${analysis.status}`,
          status: analysis.status,
        });
      }

      res.json({
        success: true,
        results: analysis.results,
        summary: analysis.summary || {
          totalOutliers: analysis.total_outliers_found,
          channelsAnalyzed: analysis.total_channels_analyzed,
          processingTime: analysis.processing_time_ms,
          config: analysis.config,
        },
      });
    } catch (error) {
      logger.error('Error fetching analysis results:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error',
      });
    }
  });

// Export results as CSV
router.get('/export/:analysisId',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res) => {
    try {
      const { analysisId } = req.params;
      const analysis = await analysisRepository.findById(analysisId);

      if (!analysis || analysis.status !== 'completed') {
        return res.status(404).json({
          success: false,
          message: 'Analysis not found or not completed',
        });
      }

      // Check ownership (users can only access their own analyses, admins can access all)
      if (analysis.user_id !== req.user.id && req.user.role !== 'admin') {
        return res.status(403).json({
          success: false,
          message: 'Access denied',
        });
      }

      // Generate CSV
      const headers = [
        'Channel Name',
        'Subscribers',
        'Video Title',
        'Views',
        'Outlier Score',
        'Brand Fit',
        'Game/Content',
        'URL',
        'Published Date',
      ];

      const csvRows = [headers.join(',')];

      if (analysis.results && analysis.results.length > 0) {
        analysis.results.forEach(video => {
          const row = [
            `"${video.channelInfo?.snippet?.title || 'N/A'}"`,
            video.channelInfo?.statistics?.subscriberCount || 0,
            `"${video.snippet?.title || 'N/A'}"`,
            video.statistics?.viewCount || 0,
            video.outlierScore?.toFixed(1) || '0.0',
            video.brandFit?.toFixed(1) || '0.0',
            `"${video.snippet?.tags ? video.snippet.tags[0] : 'N/A'}"`,
            `"https://youtube.com/watch?v=${video.id?.videoId || ''}"`,
            video.snippet?.publishedAt || '',
          ];
          csvRows.push(row.join(','));
        });
      }

      const csv = csvRows.join('\n');

      res.setHeader('Content-Type', 'text/csv');
      res.setHeader('Content-Disposition', `attachment; filename="outliers-${analysisId}.csv"`);
      res.send(csv);
    } catch (error) {
      logger.error('Error exporting analysis results:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error',
      });
    }
  });

// Export results as Excel
router.get('/export/:analysisId/excel',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res) => {
    try {
      const { analysisId } = req.params;
      const analysis = await analysisRepository.findById(analysisId);

      if (!analysis || analysis.status !== 'completed') {
        return res.status(404).json({
          success: false,
          message: 'Analysis not found or not completed',
        });
      }

      // Check ownership (users can only access their own analyses, admins can access all)
      if (analysis.user_id !== req.user.id && req.user.role !== 'admin') {
        return res.status(403).json({
          success: false,
          message: 'Access denied',
        });
      }

      // Create Excel workbook
      const workbook = new ExcelJS.Workbook();
      workbook.creator = 'YouTube Outlier Discovery Tool';
      workbook.lastModifiedBy = 'System';
      workbook.created = new Date();
      workbook.modified = new Date();

      // Summary Sheet
      const summarySheet = workbook.addWorksheet('Analysis Summary', {
        properties: { tabColor: { argb: 'FF4472C4' } }
      });

      // Add summary data
      summarySheet.columns = [
        { header: 'Metric', key: 'metric', width: 25 },
        { header: 'Value', key: 'value', width: 30 }
      ];

      const summaryData = [
        { metric: 'Analysis ID', value: analysis.id },
        { metric: 'Status', value: analysis.status },
        { metric: 'Started At', value: analysis.started_at ? new Date(analysis.started_at).toLocaleString() : 'N/A' },
        { metric: 'Completed At', value: analysis.completed_at ? new Date(analysis.completed_at).toLocaleString() : 'N/A' },
        { metric: 'Processing Time (ms)', value: analysis.processing_time_ms || 'N/A' },
        { metric: 'Total Outliers Found', value: analysis.total_outliers_found || 0 },
        { metric: 'Total Channels Analyzed', value: analysis.total_channels_analyzed || 0 },
        { metric: 'Time Window (days)', value: analysis.config?.timeWindow || 'N/A' },
        { metric: 'Min Subscribers', value: analysis.config?.minSubs || 'N/A' },
        { metric: 'Max Subscribers', value: analysis.config?.maxSubs || 'N/A' },
        { metric: 'Outlier Threshold', value: analysis.config?.outlierThreshold || 'N/A' },
        { metric: 'Exclusion Channels', value: analysis.config?.exclusionChannels?.join(', ') || 'None' },
      ];

      summarySheet.addRows(summaryData);

      // Style summary sheet header
      summarySheet.getRow(1).font = { bold: true, color: { argb: 'FFFFFFFF' } };
      summarySheet.getRow(1).fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF4472C4' } };

      // Outlier Videos Sheet
      const outliersSheet = workbook.addWorksheet('Outlier Videos', {
        properties: { tabColor: { argb: 'FF70AD47' } }
      });

      // Add outlier videos data
      outliersSheet.columns = [
        { header: 'Channel Name', key: 'channelName', width: 25 },
        { header: 'Subscribers', key: 'subscribers', width: 15 },
        { header: 'Video Title', key: 'videoTitle', width: 40 },
        { header: 'Views', key: 'views', width: 15 },
        { header: 'Outlier Score', key: 'outlierScore', width: 15 },
        { header: 'Brand Fit', key: 'brandFit', width: 12 },
        { header: 'Game/Content', key: 'content', width: 20 },
        { header: 'URL', key: 'url', width: 30 },
        { header: 'Published Date', key: 'publishedAt', width: 20 },
        { header: 'Duration', key: 'duration', width: 12 },
        { header: 'Tags', key: 'tags', width: 30 }
      ];

      if (analysis.results && analysis.results.length > 0) {
        const outlierRows = analysis.results.map(video => ({
          channelName: video.channelInfo?.snippet?.title || 'N/A',
          subscribers: video.channelInfo?.statistics?.subscriberCount || 0,
          videoTitle: video.snippet?.title || 'N/A',
          views: video.statistics?.viewCount || 0,
          outlierScore: video.outlierScore?.toFixed(1) || '0.0',
          brandFit: video.brandFit?.toFixed(1) || '0.0',
          content: video.snippet?.tags ? video.snippet.tags[0] : 'N/A',
          url: `https://youtube.com/watch?v=${video.id?.videoId || ''}`,
          publishedAt: video.snippet?.publishedAt ? new Date(video.snippet.publishedAt).toLocaleDateString() : 'N/A',
          duration: video.contentDetails?.duration || 'N/A',
          tags: video.snippet?.tags?.slice(0, 5).join(', ') || 'N/A'
        }));

        outliersSheet.addRows(outlierRows);
      }

      // Style outliers sheet header
      outliersSheet.getRow(1).font = { bold: true, color: { argb: 'FFFFFFFF' } };
      outliersSheet.getRow(1).fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF70AD47' } };

      // Channel Statistics Sheet
      const channelsSheet = workbook.addWorksheet('Channel Statistics', {
        properties: { tabColor: { argb: 'FFFFC000' } }
      });

      // Calculate channel statistics
      const channelStats = {};
      if (analysis.results && analysis.results.length > 0) {
        analysis.results.forEach(video => {
          const channelName = video.channelInfo?.snippet?.title || 'Unknown';
          if (!channelStats[channelName]) {
            channelStats[channelName] = {
              channelName,
              subscribers: video.channelInfo?.statistics?.subscriberCount || 0,
              totalVideos: 0,
              totalViews: 0,
              avgOutlierScore: 0,
              avgBrandFit: 0,
              maxOutlierScore: 0
            };
          }
          
          channelStats[channelName].totalVideos++;
          channelStats[channelName].totalViews += parseInt(video.statistics?.viewCount || 0);
          channelStats[channelName].avgOutlierScore += parseFloat(video.outlierScore || 0);
          channelStats[channelName].avgBrandFit += parseFloat(video.brandFit || 0);
          channelStats[channelName].maxOutlierScore = Math.max(
            channelStats[channelName].maxOutlierScore,
            parseFloat(video.outlierScore || 0)
          );
        });

        // Calculate averages
        Object.values(channelStats).forEach(stats => {
          stats.avgOutlierScore = (stats.avgOutlierScore / stats.totalVideos).toFixed(1);
          stats.avgBrandFit = (stats.avgBrandFit / stats.totalVideos).toFixed(1);
          stats.maxOutlierScore = stats.maxOutlierScore.toFixed(1);
        });
      }

      channelsSheet.columns = [
        { header: 'Channel Name', key: 'channelName', width: 25 },
        { header: 'Subscribers', key: 'subscribers', width: 15 },
        { header: 'Outlier Videos Found', key: 'totalVideos', width: 20 },
        { header: 'Total Views', key: 'totalViews', width: 15 },
        { header: 'Avg Outlier Score', key: 'avgOutlierScore', width: 18 },
        { header: 'Max Outlier Score', key: 'maxOutlierScore', width: 18 },
        { header: 'Avg Brand Fit', key: 'avgBrandFit', width: 15 }
      ];

      if (Object.keys(channelStats).length > 0) {
        channelsSheet.addRows(Object.values(channelStats));
      }

      // Style channels sheet header
      channelsSheet.getRow(1).font = { bold: true, color: { argb: 'FFFFFFFF' } };
      channelsSheet.getRow(1).fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFC000' } };

      // Add auto-filter to all sheets
      summarySheet.autoFilter = {
        from: 'A1',
        to: 'B' + (summaryData.length + 1)
      };

      if (analysis.results && analysis.results.length > 0) {
        outliersSheet.autoFilter = {
          from: 'A1',
          to: 'K' + (analysis.results.length + 1)
        };

        if (Object.keys(channelStats).length > 0) {
          channelsSheet.autoFilter = {
            from: 'A1',
            to: 'G' + (Object.keys(channelStats).length + 1)
          };
        }
      }

      // Set response headers
      res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
      res.setHeader('Content-Disposition', `attachment; filename="outliers-analysis-${analysisId}.xlsx"`);

      // Write to response
      await workbook.xlsx.write(res);
      res.end();

    } catch (error) {
      logger.error('Error exporting analysis results as Excel:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error',
      });
    }
  });

// List all analyses
router.get('/list',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res) => {
    try {
      // Filter analyses by user ownership (admins can see all)
      const analyses = await analysisRepository.findByUserId(
        req.user.role === 'admin' ? null : req.user.id
      );

      const userAnalyses = analyses.map(analysis => ({
        id: analysis.id,
        userId: analysis.user_id, // Include userId for admin view
        status: analysis.status,
        startTime: analysis.started_at,
        endTime: analysis.completed_at,
        resultCount: analysis.results ? analysis.results.length : 0,
        config: {
          exclusionChannels: analysis.config?.exclusionChannels || [],
          timeWindow: analysis.config?.timeWindow || 7,
        },
      }));

      res.json({
        success: true,
        analyses: userAnalyses.sort((a, b) => new Date(b.startTime) - new Date(a.startTime)),
      });
    } catch (error) {
      logger.error('Error fetching analyses list:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error',
      });
    }
  });

module.exports = router;

File: server-python/src/routes/outlier.py
from flask import Blueprint, request, jsonify, g
from services.outlier_detection_service import OutlierDetectionService
from utils.logger import logger
from middleware.auth import authenticate
from middleware.rbac import require_permission, require_scopes
import uuid
import json
from db import redis_client

bp = Blueprint('outlier', __name__, url_prefix='/api/outlier')
outlier_service = OutlierDetectionService()  # Initialize without socketio

def init_outlier_service(socketio_instance=None):
    global outlier_service
    # For now, we're not using socketio, so we don't need to initialize it
    pass

@bp.route('/start', methods=['POST'])
@authenticate
@require_permission('analysis:write')
@require_scopes(['write'])
def start_analysis():
    """
    Start new outlier analysis
    """
    data = request.get_json()
    if not data:
        return jsonify({'success': False, 'error': 'No data provided'}), 400
    
    # Validate required fields
    required_fields = ['exclusionChannels', 'minSubs', 'maxSubs', 'timeWindow', 'outlierThreshold']
    for field in required_fields:
        if field not in data:
            return jsonify({'success': False, 'error': f'Missing required field: {field}'}), 400
    
    # Validate data types
    if not isinstance(data['exclusionChannels'], list):
        return jsonify({'success': False, 'error': 'Exclusion channels must be an array'}), 400
    
    if not isinstance(data['minSubs'], int) or data['minSubs'] < 1000:
        return jsonify({'success': False, 'error': 'Minimum subscribers must be at least 1000'}), 400
    
    if not isinstance(data['maxSubs'], int) or data['maxSubs'] < 10000:
        return jsonify({'success': False, 'error': 'Maximum subscribers must be at least 10000'}), 400
    
    if not isinstance(data['timeWindow'], int) or not (1 <= data['timeWindow'] <= 30):
        return jsonify({'success': False, 'error': 'Time window must be between 1-30 days'}), 400
    
    if not isinstance(data['outlierThreshold'], int) or not (10 <= data['outlierThreshold'] <= 100):
        return jsonify({'success': False, 'error': 'Outlier threshold must be between 10-100'}), 400
    
    try:
        analysis_id = str(uuid.uuid4())
        config = data
        user_id = getattr(g, 'user', {}).get('id')
        
        # Store analysis record in Redis (in production, use a real database)
        analysis_data = {
            'id': analysis_id,
            'user_id': user_id,
            'name': config.get('name', f'Analysis {analysis_id}'),
            'description': config.get('description'),
            'config': config,
            'status': 'pending',
            'created_at': __import__('datetime').datetime.utcnow().isoformat()
        }
        
        redis_client.setex(f'analysis:{analysis_id}', 86400, json.dumps(analysis_data))
        
        # Start analysis in background
        import threading
        def run_analysis():
            try:
                outlier_service.start_analysis(analysis_id, config)
            except Exception as e:
                logger.error(f"Analysis {analysis_id} failed: {e}")
                # Update status to failed
                analysis_data['status'] = 'failed'
                analysis_data['error_message'] = str(e)
                redis_client.setex(f'analysis:{analysis_id}', 86400, json.dumps(analysis_data))
        
        thread = threading.Thread(target=run_analysis)
        thread.start()
        
        return jsonify({
            'success': True,
            'analysisId': analysis_id,
            'message': 'Analysis started successfully'
        }), 200
        
    except Exception as e:
        logger.error(f"Error starting analysis: {e}")
        return jsonify({'success': False, 'error': 'Failed to start analysis'}), 500

@bp.route('/status/<string:analysis_id>', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def get_analysis_status(analysis_id):
    """
    Get analysis status
    """
    try:
        # Get analysis from Redis
        analysis_data = redis_client.get(f'analysis:{analysis_id}')
        if not analysis_data:
            return jsonify({'success': False, 'error': 'Analysis not found'}), 404
        
        analysis = json.loads(analysis_data)
        
        # Check ownership (users can only access their own analyses, admins can access all)
        user = getattr(g, 'user', {})
        if analysis.get('user_id') != user.get('id') and user.get('role') != 'admin':
            return jsonify({'success': False, 'error': 'Access denied'}), 403
        
        # Get real-time status from outlier service
        status = outlier_service.get_analysis_status(analysis_id)
        
        return jsonify({
            'success': True,
            'analysis': {
                'id': analysis['id'],
                'status': analysis.get('status', 'unknown'),
                'started_at': analysis.get('created_at'),
                'config': analysis.get('config'),
                'progress': status
            }
        }), 200
        
    except Exception as e:
        logger.error(f"Error fetching analysis status: {e}")
        return jsonify({'success': False, 'error': 'Internal server error'}), 500

@bp.route('/results/<string:analysis_id>', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def get_analysis_results(analysis_id):
    """
    Get analysis results
    """
    try:
        # Get analysis from Redis
        analysis_data = redis_client.get(f'analysis:{analysis_id}')
        if not analysis_data:
            return jsonify({'success': False, 'error': 'Analysis not found'}), 404
        
        analysis = json.loads(analysis_data)
        
        # Check ownership (users can only access their own analyses, admins can access all)
        user = getattr(g, 'user', {})
        if analysis.get('user_id') != user.get('id') and user.get('role') != 'admin':
            return jsonify({'success': False, 'error': 'Access denied'}), 403
        
        # Check if analysis is completed
        if analysis.get('status') != 'completed':
            return jsonify({
                'success': False, 
                'error': f'Analysis is {analysis.get("status", "in progress")}',
                'status': analysis.get('status', 'in progress')
            }), 400
        
        # Get results from Redis
        results_data = redis_client.get(f'analysis_results:{analysis_id}')
        results = json.loads(results_data) if results_data else []
        
        return jsonify({
            'success': True,
            'results': results,
            'summary': analysis.get('summary', {
                'totalOutliers': len(results) if results else 0,
                'channelsAnalyzed': 0,
                'config': analysis.get('config')
            })
        }), 200
        
    except Exception as e:
        logger.error(f"Error fetching analysis results: {e}")
        return jsonify({'success': False, 'error': 'Internal server error'}), 500

@bp.route('/list', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def list_analyses():
    """
    List all analyses for the user
    """
    try:
        user = getattr(g, 'user', {})
        user_id = user.get('id')
        is_admin = user.get('role') == 'admin'
        
        # In a real implementation, this would query a database
        # For now, we'll return an empty list
        analyses = []
        
        user_analyses = []
        for analysis in analyses:
            # Filter analyses by user ownership (admins can see all)
            if is_admin or analysis.get('user_id') == user_id:
                user_analyses.append({
                    'id': analysis['id'],
                    'userId': analysis.get('user_id'),
                    'status': analysis.get('status'),
                    'startTime': analysis.get('created_at'),
                    'resultCount': len(analysis.get('results', [])),
                    'config': {
                        'exclusionChannels': analysis.get('config', {}).get('exclusionChannels', []),
                        'timeWindow': analysis.get('config', {}).get('timeWindow', 7)
                    }
                })
        
        # Sort by start time (newest first)
        user_analyses.sort(key=lambda x: x['startTime'], reverse=True)
        
        return jsonify({
            'success': True,
            'analyses': user_analyses
        }), 200
        
    except Exception as e:
        logger.error(f"Error fetching analyses list: {e}")
        return jsonify({'success': False, 'error': 'Internal server error'}), 500

@bp.route('/export/<string:analysis_id>/<string:format_type>', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def export_analysis_legacy(analysis_id, format_type):
    """
    Legacy export endpoint - redirects to new export service
    """
    from services.export_service import ExportService
    from flask import Response
    from datetime import datetime
    
    export_service = ExportService()
    supported_formats = ['excel', 'csv', 'pdf', 'json', 'html']
    
    if format_type not in supported_formats:
        return jsonify({
            'success': False, 
            'error': f'Unsupported format. Supported formats: {", ".join(supported_formats)}'
        }), 400
    
    try:
        user_id = getattr(g, 'user', {}).get('id')
        
        # Check if user has access to this analysis
        analysis_data = redis_client.get(f'analysis:{analysis_id}')
        if not analysis_data:
            return jsonify({'success': False, 'error': 'Analysis not found'}), 404
        
        analysis = json.loads(analysis_data)
        user = getattr(g, 'user', {})
        if analysis.get('user_id') != user.get('id') and user.get('role') != 'admin':
            return jsonify({'success': False, 'error': 'Access denied'}), 403
        
        # Check if analysis is completed
        if analysis.get('status') != 'completed':
            return jsonify({
                'success': False, 
                'error': f'Analysis is {analysis.get("status", "in progress")}. Cannot export incomplete analysis.'
            }), 400
        
        # Generate export
        try:
            if format_type == 'excel':
                result = export_service.export_to_excel(analysis_id, user_id)
                mimetype = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
                filename = f'analysis_{analysis_id}_{datetime.now().strftime("%Y%m%d_%H%M%S")}.xlsx'
                
            elif format_type == 'csv':
                result = export_service.export_to_csv(analysis_id, user_id)
                mimetype = 'text/csv'
                filename = f'analysis_{analysis_id}_{datetime.now().strftime("%Y%m%d_%H%M%S")}.csv'
                
            elif format_type == 'pdf':
                result = export_service.export_to_pdf(analysis_id, user_id)
                mimetype = 'application/pdf'
                filename = f'analysis_{analysis_id}_{datetime.now().strftime("%Y%m%d_%H%M%S")}.pdf'
                
            elif format_type == 'json':
                result = export_service.export_to_json(analysis_id, user_id)
                mimetype = 'application/json'
                filename = f'analysis_{analysis_id}_{datetime.now().strftime("%Y%m%d_%H%M%S")}.json'
                
            elif format_type == 'html':
                result = export_service.export_to_html(analysis_id, user_id)
                mimetype = 'text/html'
                filename = f'analysis_{analysis_id}_{datetime.now().strftime("%Y%m%d_%H%M%S")}.html'
            
            # Return file response
            if isinstance(result, str):
                # String content (JSON, HTML)
                return Response(
                    result,
                    mimetype=mimetype,
                    headers={
                        'Content-Disposition': f'attachment; filename="{filename}"',
                        'Content-Type': mimetype
                    }
                )
            else:
                # Binary content (Excel, PDF) or StringIO (CSV)
                return Response(
                    result.getvalue(),
                    mimetype=mimetype,
                    headers={
                        'Content-Disposition': f'attachment; filename="{filename}"',
                        'Content-Type': mimetype
                    }
                )
                
        except ValueError as e:
            return jsonify({'success': False, 'error': str(e)}), 400
        except Exception as e:
            logger.error(f"Export generation failed for {analysis_id}, format {format_type}: {e}")
            return jsonify({'success': False, 'error': 'Export generation failed'}), 500
        
    except Exception as e:
        logger.error(f"Error exporting analysis {analysis_id}: {e}")
        return jsonify({'success': False, 'error': 'Internal server error'}), 500
